File: android/app/src/test/java/ai/oriveo/community/core/util/GraphemeUtilsTest.kt
package ai.oriveo.community.core.util

import org.junit.Assert.assertEquals
import org.junit.Test

class GraphemeUtilsTest {

    @Test
    fun `graphemeCount keeps emoji and combining marks intact`() {
        assertEquals(1, "😀".graphemeCount())
        assertEquals(1, "e\u0301".graphemeCount())
        assertEquals(1, "👍🏽".graphemeCount())
        assertEquals(3, "A😀B".graphemeCount())
    }

    @Test
    fun `takeGraphemes does not split grapheme clusters`() {
        assertEquals("😀", "😀x".takeGraphemes(1))
        assertEquals("e\u0301", "e\u0301x".takeGraphemes(1))
        assertEquals("👍🏽", "👍🏽x".takeGraphemes(1))
    }

    // -- extended grapheme tests --

    @Test
    fun `family emoji counts as 1 grapheme`() {
        // 👨‍👩‍👧‍👦 = U+1F468 U+200D U+1F469 U+200D U+1F467 U+200D U+1F466
        assertEquals(1, "\uD83D\uDC68\u200D\uD83D\uDC69\u200D\uD83D\uDC67\u200D\uD83D\uDC66".graphemeCount())
    }

    @Test
    fun `flag emoji counts as 1 grapheme`() {
        // 🇨🇳 = U+1F1E8 U+1F1F3
        assertEquals(1, "\uD83C\uDDE8\uD83C\uDDF3".graphemeCount())
    }

    @Test
    fun `combining character e-acute counts as 1 grapheme`() {
        // e + combining acute accent = é
        assertEquals(1, "e\u0301".graphemeCount())
    }

    @Test
    fun `takeGraphemes at boundary does not split family emoji`() {
        val familyEmoji = "\uD83D\uDC68\u200D\uD83D\uDC69\u200D\uD83D\uDC67\u200D\uD83D\uDC66"
        val text = "${familyEmoji}ABC"
        // Taking 1 grapheme should keep the family emoji intact
        assertEquals(familyEmoji, text.takeGraphemes(1))
        // Taking 2 graphemes should include the family emoji plus A
        assertEquals("${familyEmoji}A", text.takeGraphemes(2))
    }

    @Test
    fun `takeGraphemes at boundary does not split flag emoji`() {
        val flag = "\uD83C\uDDE8\uD83C\uDDF3"
        val text = "${flag}Hi"
        assertEquals(flag, text.takeGraphemes(1))
        assertEquals("${flag}H", text.takeGraphemes(2))
    }

    @Test
    fun `takeGraphemes 2000 with mixed emoji and text`() {
        // Build a mixed string of emoji and plain text well beyond 2000 graphemes
        val emoji = "😀"
        val segment = "Hello${emoji}" // 6 graphemes
        val repeats = 400 // 2400 graphemes
        val text = segment.repeat(repeats)
        val result = text.takeGraphemes(2000)
        assertEquals(2000, result.graphemeCount())
    }

    @Test
    fun `takeGraphemes with zero returns empty string`() {
        assertEquals("", "Hello".takeGraphemes(0))
    }

    @Test
    fun `takeGraphemes with negative returns empty string`() {
        assertEquals("", "Hello".takeGraphemes(-1))
    }

    @Test
    fun `takeGraphemes on empty string returns empty`() {
        assertEquals("", "".takeGraphemes(5))
    }

    @Test
    fun `takeGraphemes when limit exceeds length returns full string`() {
        assertEquals("ABC", "ABC".takeGraphemes(100))
    }

    @Test
    fun `graphemeCount for mixed CJK and emoji`() {
        // kana characters + emoji + ASCII
        val text = "ねこ😀AB"
        assertEquals(5, text.graphemeCount())
    }

    @Test
    fun `graphemeCount for empty string is 0`() {
        assertEquals(0, "".graphemeCount())
    }
}

File: web/apps/app/lib/utils/__tests__/grapheme-utils.test.ts
/**
 * Unit tests for grapheme-utils.
 *
 * Covers grapheme cluster counting and truncation built on Intl.Segmenter, focusing on Unicode
 * edge cases such as emoji, combining marks and flags, plus the fallback path for browsers with
 * no Intl.Segmenter (Firefox before 125).
 */
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { graphemeCount, takeGraphemes } from '../grapheme-utils';

describe('graphemeCount', () => {
  it('counts ASCII strings by character', () => {
    expect(graphemeCount('hello')).toBe(5);
    expect(graphemeCount('abc123')).toBe(6);
  });

  it('returns 0 for an empty string', () => {
    expect(graphemeCount('')).toBe(0);
  });

  it('counts a family emoji (ZWJ sequence) as one grapheme', () => {
    // 👨‍👩‍👧‍👦 = U+1F468 U+200D U+1F469 U+200D U+1F467 U+200D U+1F466
    expect(graphemeCount('👨‍👩‍👧‍👦')).toBe(1);
  });

  it('counts a flag emoji (regional indicator) as one grapheme', () => {
    // 🇨🇳 = U+1F1E8 U+1F1F3
    expect(graphemeCount('🇨🇳')).toBe(1);
  });

  it('counts a combining mark as one grapheme', () => {
    // é = e + U+0301 (combining acute accent)
    const eCombined = 'e\u0301';
    expect(graphemeCount(eCombined)).toBe(1);
  });

  it('counts mixed content correctly', () => {
    // "Hi👨‍👩‍👧‍👦!" = H + i + 👨‍👩‍👧‍👦 + ! = 4
    expect(graphemeCount('Hi👨‍👩‍👧‍👦!')).toBe(4);
  });

  it('counts each of several emoji as one', () => {
    expect(graphemeCount('🇨🇳🇯🇵🇺🇸')).toBe(3);
  });
});

describe('takeGraphemes', () => {
  it('truncation does not break ASCII characters', () => {
    expect(takeGraphemes('hello world', 5)).toBe('hello');
  });

  it('returns the whole string when limit exceeds its length', () => {
    expect(takeGraphemes('short', 100)).toBe('short');
  });

  it('returns an empty string when limit is 0', () => {
    expect(takeGraphemes('hello', 0)).toBe('');
  });

  it('truncating on an emoji boundary does not break a ZWJ sequence', () => {
    // A family emoji is one grapheme, so limit=1 returns the whole emoji
    const family = '👨‍👩‍👧‍👦';
    expect(takeGraphemes(`${family}abc`, 1)).toBe(family);
  });

  it('truncating on a flag emoji boundary does not break the regional indicator pair', () => {
    expect(takeGraphemes('🇨🇳abc', 1)).toBe('🇨🇳');
  });

  it('truncates mixed content on grapheme boundaries', () => {
    // "Hi👨‍👩‍👧‍👦!" → limit=3 → "Hi👨‍👩‍👧‍👦"
    expect(takeGraphemes('Hi👨‍👩‍👧‍👦!', 3)).toBe('Hi👨‍👩‍👧‍👦');
  });

  it('returns an empty string for an empty string', () => {
    expect(takeGraphemes('', 5)).toBe('');
  });
});

describe('fallback path when Intl.Segmenter is unavailable (Firefox before 125)', () => {
  // Delegate through the prototype to keep the rest of Intl, shadowing only Segmenter as undefined to emulate an old browser
  const intlWithoutSegmenter = Object.create(Intl, {
    Segmenter: { value: undefined },
  }) as typeof Intl;

  beforeEach(() => {
    vi.resetModules();
    vi.stubGlobal('Intl', intlWithoutSegmenter);
  });

  afterEach(() => {
    vi.unstubAllGlobals();
    vi.resetModules();
  });

  function importFresh() {
    return import('../grapheme-utils');
  }

  it('evaluating the module top level does not throw: a top-level new Intl.Segmenter blows up as soon as the chunk is evaluated', async () => {
    await expect(importFresh()).resolves.toBeDefined();
  });

  it('graphemeCount falls back to counting code points', async () => {
    const { graphemeCount: count } = await importFresh();
    expect(count('hello')).toBe(5);
    expect(count('')).toBe(0);
    // Fallback semantics: a ZWJ family emoji is 4 emoji code points plus 3 ZWJ = 7
    expect(count('👨‍👩‍👧‍👦')).toBe(7);
    // A flag is 2 regional indicator code points
    expect(count('🇨🇳')).toBe(2);
    // e plus a combining accent is 2 code points
    const eCombined = 'e' + String.fromCharCode(0x0301);
    expect(count(eCombined)).toBe(2);
  });

  it('takeGraphemes falls back to code point truncation without splitting a surrogate pair', async () => {
    const { takeGraphemes: take } = await importFresh();
    expect(take('hello world', 5)).toBe('hello');
    expect(take('hello', 0)).toBe('');
    expect(take('short', 100)).toBe('short');
    // A grinning face is a single code point (a surrogate pair), and Array.from does not cut it in half
    expect(take('😀abc', 1)).toBe('😀');
    expect(take('😀abc', 2)).toBe('😀a');
  });

  it('a negative limit returns an empty string, matching the Segmenter path', async () => {
    const { takeGraphemes: take } = await importFresh();
    expect(take('hello', -1)).toBe('');
  });
});
